fix: update decline in a simple way

reliveyy opened the pull request:
This PR implement a very simple fix for update decline by just bypassing ensuring all services are up and running and any post actions. If the user declines the update he will enter into the xud-ctl shell directly.

The edge cases:
1. If the new setup removes some image and it is stopped or removed in current environment, then there is no way to bring it back automatically. And on the other side, if the new setup removes some service but the old setup keep it running, then this service needs to manually stop and remove.
2. If your setup is up-to-date then you will use the new utils (setup) to ensure all services are up and do post actions. This could be error prone.

**UPDATE**

A significant change was made in "preview! fix false update (down/up case)" which introduced a new concept "snapshot" in xud-docker. The updates were calculated by differentiating two snapshots since this commit. And the next important fix was "fix setup.sh" which brings the utils into *update checking* stage. After these two major fixes, the above two edge cases were resolved. But it has changed some fundamental logic in xud-docker. So it took so much time to do regression testing. And even now I cannot make myself believe all cases are covered.

So in order to finish this PR soon, we got two options

1) Revoke all changes since "preview! fix false update (down/up case)" and accept an imperfect solution with some failing cases in @raladev comments

> environment up-to-date, one container was stopped (connext) -> connext still is stopped VS connext up w/o any question --- FAILED

> environment up-to-date, all containers stopped (except of xud) -> all containers still are stopped VS they up without any question --- FAILED

> environment up-to-date, one container was stopped (xud) -> xud still is stopped VS question about ports update --- FAILED

This one was fixed by commit "fix ports diff"

2) Keep going with current "perfect" solution and it definitely needs more time to make it stable.

reliveyy commented:
> Something wrong here

All containers missing case criterion has been narrowed down to avoid auto-update for existing environment. @kilrau

kilrau commented:
Update:
- Case 4: **passed**
- Case 5: **failed**, shouldn't bring update prompt and just create and start containers since they are **missing**, meaning the environment is down and not out-of-date. This is wrong behaviour we have since ever, let's fix it now since it avoids above failure.

reliveyy commented:
What has been changed in order to meet testing case 4 and 5:

**Before**
If all containers are missing then auto-update.

**After**
If all containers are mssing:
1. Newly installed environment: auto-update
2. Existing environment & has image update: auto-update (Testing case 4)
3. Existing environment & no image update: NO update (Testing case 5) -> _This is the main difference_

Potential problems introduced by testing case 5 fix:
1. If you shutdown your environment then you cannot update (It will bring up the environment directly. In another word, if you want to update your "down" environment you need to run xud script 2 times);
2. You could get unintended container update when you has a "down" environment (They are created by new utils w/ new configurations probably).

raladev commented:
My results:
- environment up and healthy + new versions & decline -> nothing changed  --- PASSED
- environment up and healthy + new versions & accept -> containers updated --- PASSED

- environment up and healthy + new versions + remove arby from cmd line & decline -> nothing changed --- PASSED
- environment up and healthy + new versions + add arby to cmd line (prev installation has no arby) & decline -> nothing changed --- PASSED
- environment up and healthy + new versions + remove arby from cmd line & decline -> arby was removed ---- PASSED

- environment up, one container was manually removed (arby) + new version avalible (arby) & decline -> arby still is down --- PASSED 
- environment up, one container was manually removed (arby) + new version avalible (arby) & accept -> arby is updated and up --- PASSED

- environment up, one container was manually removed (arby) & decline -> arby still is down --- PASSED
- environment up, one container was manually removed (arby) & accept -> arby up (Note: got freeze on update checking once, may be network problem) --- PASSED

- environment up, one container was stopped (connext) + new version avalible (connext) & decline -> connext still is stopped --- PASSED 
- environment up, one container was stopped (connext) + new version avalible (connext) & accept -> connext is updated and up --- PASSED
- env with old images, all containers stopped & decline -> containers still are down -- PASSED

- environment up-to-date, one container was stopped (connext) -> connext still is stopped VS connext up w/o any question --- FAILED
- environment up-to-date, all containers stopped (except of xud) ->  all containers still are stopped VS they up without any question --- FAILED
- environment up-to-date, one container was stopped (xud) -> xud still is stopped VS question about ports update --- FAILED

- environment was FULLY down with available updates & decline -> No asking about updates, just new versions was downloaded for several images --- PASSED but dont like it, i want old images with question about update

- environment up + up-to-date -> just enter to cli --- PASSED
- environment down + up-to-date -> just enter to cli and create containers --- PASSED

reliveyy commented:
The recent changes fixed "down" use case which has false updates available. We export docker-compose yaml like snaphost file into `network_dir/logs/snapshots/snapshot-timestamp.yml` to persist your running environment in case you "down" the whole environment (which will remove all containers). And we are using snapshots diff to detect updates now.

BTW, the update details is reformated to **service** perspective. @kilrau @raladev

raladev commented:
upd 2 is a branch with different xud/arby versions.

- no env, clean installation with empty simnet folder --- FAILED --- asking about update for n/a images.
**Expected result:** No update asking for clean installation

- environment up, one container was manually removed (arby) & decline -> arby still is down  --- FAILED ---

**Steps:**
1. bash xud.sh -b fix-update-decline --arby.disabled false
2. bash xud.sh -b fix-update-decline  -> decline update

**Actual Result:** utils tries to start arby, when i dont want to start it + python err after second No (and same for cases with arby updating from downed state)

**Expected result:** no arby because of update decline.
[Screenshot from 2020-08-10 17-35-15](https://user-images.githubusercontent.com/29906866/89796602-84bc4480-db32-11ea-91b6-97c6c62fdab2.png)

- environment was FULLY down with available updates & decline -> No asking about updates, just new versions was downloaded for several images --- FAILED --- after no answer utils does not start old setup

**Steps:**
1. bash xud.sh -b upd2 --arby.disabled false
2. down in cli console
3. bash xud.sh -b fix-update-decline --arby.disabled false  -> decline update

**Actual result:**
Services errors + python err

**Expected result:**
old env is up (upd2)
[Screenshot from 2020-08-10 17-49-02](https://user-images.githubusercontent.com/29906866/89797343-7de20180-db33-11ea-8c24-0fe390c6e5e2.png)

- environment up and healthy + new versions + remove arby from cmd line & decline -> nothing changed --- FAILED --- no arby in status command, but container works, also after second run there is no question about arby, but it is still online.

**Steps:**
1. bash xud.sh -b upd2 --arby.disabled false
2. bash xud.sh -b fix-update-decline  -> decline update
4. check status -> no arby
5. check docker ps -a -> arby is online

**Expected result**
No update for arby (it works) + arby is in status list (it does not work)

- environment up and healthy + new versions + add arby to cmd line (prev installation has no arby) & decline -> nothing changed --- FAILED --- arby is in status command, but there is no arby container
_Like previous point but vice versa_

- case when user have both images and runned env --- FAILED --- failed because utils says All up to date, but xud and arby should be recreated with correct images

**Steps:**
1. bash xud.sh -b upd2 --arby.disabled false
2. bash xud.sh -b fix-update-decline --arby.disabled false  -> accept update
3. bash xud.sh -b upd2 --arby.disabled false

**Actual result:**
No update asking, All-up-to-date

**Expected result:**
asking about xud and arby update.

raladev commented:
i did not test it fully because common cases contains strange misbehavior:

- [ ] clean installation, with empty simnet folder w/o containers and images ---FAILED---

Steps: 
1. bash xud.sh -b upd3

Actual result: 
1. question about update
2. `docker images` contains 8 images instead of 4 and latest image has same hash as _upd3
3. `docker ps -a` - latest images are used instead of _upd3

Expected result: 
1. NO question about update
2. `docker images` contains only 4 images with _upd3 suffix

- [ ] environment up and healthy + new versions are available --- FAILED

Steps: 
1. bash xud.sh -b upd3 (hav different xud/connext/arby hashes)
2. bash xud.sh -b fix-update-decline

Actual result: 
1. NO question about update, force update

Expected result: 
1. question about update is appeared

reliveyy commented:
@raladev 

### Case 1. clean installation, with empty simnet folder w/o containers and images

Expected result:

> NO question about update

Fixed by https://github.com/ExchangeUnion/xud-docker/pull/639/commits/9f0e4603409e59231c4f7623bd393cef4bb9d993

> docker images contains only 4 images with _upd3 suffix

Now bring back the old branch image implementation https://github.com/ExchangeUnion/xud-docker/pull/639/commits/3fc24ab5050f682d73da868e31ddc7ed885fd6c5 (I've dropped it out becuase it will save some logic but it's not clear when using`docker ps` to tell which image uses branch version)

### Case 2. environment up and healthy + new versions are available

Expected result:

> question about update is appeared

Should be fixed by the same commit above. https://github.com/ExchangeUnion/xud-docker/pull/639/commits/9f0e4603409e59231c4f7623bd393cef4bb9d993

raladev commented:
--- PASSED --- no env -> instllation without asking 
--- PASSED --- environment up and healthy + new versions & accept -> containers updated --- PASSED

--- FAILED --- environment up and healthy + new versions & decline -> nothing changed --- PASSED
--- FAILED --- environment down + up-to-date -> just enter to cli and create containers --- PASSED
--- FAILED --- environment up + up-to-date -> just enter to cli --- PASSED

- [ ] all three cases were failed because of incorrect utils image pulling

Steps:
1. bash xud.sh -b upd4
2. exit in cli
3. bash xud.sh -b upd4

Actual result:
latest utils pulling + err in console

Expected result: 
no utils pulling because i have required utils image 

```
1) Simnet
2) Testnet
3) Mainnet
Please choose the network: 1
🚀 Launching simnet environment
🌍 Checking for updates...
Unable to find image 'exchangeunion/utils:latest' locally
latest: Pulling from exchangeunion/utils
cbdbe7a5bc2a: Already exists 
26ebcd19a4e3: Pull complete 
35acdcbeccf1: Pull complete 
647f327f91dd: Pull complete 
9add0a120de2: Pull complete 
3c09ec2adc1f: Pull complete 
6793df4cd7b3: Pull complete 
f05c47042710: Pull complete 
Digest: sha256:449984152e9cd1e79a887855bbd4235698c2d6e8fe5f88a87e4e7f2214b20647
Status: Downloaded newer image for exchangeunion/utils:latest
docker: Error response from daemon: OCI runtime create failed: container_linux.go:349: starting container process caused "exec: \"-b\": executable file not found in $PATH": unknown.
```

raladev commented:
--- PASSED--- no env -> instillation without asking 
--- PASSED --- environment up + up-to-date -> just enter to cli 
--- PASSED --- environment down + up-to-date -> just enter to cli and create containers 
--- PASSED --- (but without pulling output) --- environment up + up-to-date + add new image in cmd (arby) + decline -> nothing changed

--- FAILED --- [ ] environment up + up-to-date + add new image in cmd (arby) + decline -> nothing changed --- Arby was created after No answer + I dont see arby pulling in the utils output)

Steps:
1. bash xud.sh -b upd5
2. exit in cli
3. bash xud.sh -b upd5 --arby.disable false -> Question about arby missing
4. Answer No

Actual result:
1. Arby is created after No answer
2. No arby image pulling info in utils output
![Screenshot from 2020-08-18 16-07-02](https://user-images.githubusercontent.com/29906866/90516583-e0a74e80-e16c-11ea-872f-8e8a5b90c258.png)


Expected result:
1. No arby after No answer
2. arby image pulling info is in utils output

--- FAILED --- [ ] environment up and healthy + new versions & decline -> nothing changed --- (New images were pulled and used after No answer) 
Steps:
1. bash xud.sh -b upd5 --arby.disabled false
2. exit in cli
3. bash xud.sh -b fix-update-decline --arby.disabled false -> Question about update missing
4. Answer No

Actual result:
1. I got info about new services only after update question
2. After No answer I got updates and containers recreation
3. Why i need second question about image pulling
4. No image pulling info in utils output
[Screenshot from 2020-08-18 15-49-15](https://user-images.githubusercontent.com/29906866/90516533-ccfbe800-e16c-11ea-9058-9979518c5e78.png)

Expected result:
1. Info about new services before update question (like it was with arby in previous case)
2. No pulling and update after No answer
4. image pulling info is in utils output

reliveyy commented:
- [ ] "down" command stucks without any logs
- [ ] `bash xud.sh -b fix-update-decline` throws an exception
```
❯ ~/xud.sh -b fix-update-decline                      
1) Simnet
2) Testnet
3) Mainnet
Please choose the network: 1
🚀 Launching simnet environment
🌍 Checking for updates...
- Service lndbtc: missing
- Service lndltc: missing
- Service connext: missing
  * image: new version available (exchangeunion/connext:latest__fix-update-decline, 9a360 -> 1e244)
- Service xud: missing
- Service arby: disabled
Would you like to pull new images? (Warning: this may restart your environment and cancel all open orders) [Y/n] 
💿 Pulling image exchangeunion/connext:latest__fix-update-decline...
cbdbe7a5bc2a: Already exists
091983d60f65: Already exists
7bacc183740f: Already exists
c4a5a5d270d4: Already exists
42b49ac6076c: Pull complete
a74c2a6d2da6: Pull complete
87ca68b22f98: Pull complete
3b80ecc723a0: Pull complete
7d4359ed4da0: Pull complete
📦 Creating container simnet_lndbtc_1...
📦 Creating container simnet_lndltc_1...
📦 Creating container simnet_connext_1...
📦 Creating container simnet_xud_1...
📦 Stopping container simnet_arby_1...
Traceback (most recent call last):
  File "/root/launcher/node/UpdateManager.py", line 402, in _apply
    c.stop()
AttributeError: 'NoneType' object has no attribute 'stop'
⚠️ Failed to update service arby

🏃 Warming up...


Enter master xud password:
```

reliveyy commented:
This PR branch has been revoked to the commit before we introduced snapshots. @raladev
